### DISCUSSION TITLE
osu!catch issues to address for lazer ranked (gathered from community)

### DISCUSSION THREAD

**Discussion by bastoo0** [ppy/osu#26145] [Q&A]
I went through the feedback of the osu!catch community, asking "what needs to be adressed before lazer leaderboards are switched on?". This is a very late feedback and I apologize for that. I believe people were not interested in trying stuff on lazer and reporting issues before it received an "official release date". So this is a summary of those:

### **High priority** (requires leaderboards resets if implemented):

- **Bug fix** : Notes can appear out of bounds (you can place out of bounds fruits in editor: https://github.com/ppy/osu/issues/23810), any out of bounds object will appear during gameplay. On stable, their positions are reset at x=0 or something of that effect.
This is high priority because some ranked maps are broken because of this issue (Quoting Secre: " https://osu.ppy.sh/beatmapsets/1062804#fruits/2234731 slider at 2:00 on PH diff, several other ranked or loved maps are touched ").

- **Feature implementation** : Remove object randomization for HardRock. This is the most requested and agreed upon feature change. Currently, objects receive a random offset when HR is on. This could be a fun idea but it naturally makes many patterns barely playable on HR and is a struggle for HR players.
The randomization should still exist in classic (for compatibility or just people who want it). We have been discussing keeping the randomization feature but as a future adjustable mod.
This only requires a few lines of code and pretty much no testing (I think?). **We still need to get official community feedback for that**.

- **Bug fix** : There is an existing issue on both lazer and stable where droplets are generated in an inconsistent manner on low BPM (https://github.com/ppy/osu-stable-issues/issues/24, lazer issue: https://github.com/ppy/osu/issues/6184). If we want to fix this, this is probably the right time.

- **Rebalancing** : This lazer scoring issue should probably be reviewed as it seems to create imbalance in some leaderboards: https://github.com/ppy/osu/discussions/26040 

### Medium priority (can affect gameplay in a really bad way but does not break leaderboards):

- **Bug fix** : Inconsistent layering (see https://github.com/ppy/osu/issues/25827, also not described here: droplets can appear in front of normal objects). This gives a really awkward playing experience and will probably be something that prevents many players from giving lazer a chance.

- **Rebalancing** : HP is perceived as "too punishing" in catch. On high difficulties, missing an hyperdash leading to a stack will cause an inevitable fail most of the time. HP calculation should prevent failing on a singular chunk of objects, especially at high BPM.

### Low priority (are probably out of scope here but I think should be mentioned, just so people in the community are happy, _feel free to ignore_):

- **Feature implementation** : Tiny droplets Hyperdashes: This is a stable issue. Some tiny droplets, can get in awkward positions that prevent a map from being SSable. Big droplets and normal objects are generating an hyperdash in the cases where they are too far from the next fruits, but not tiny droplets. The issue mainly appears on converts as it would prevent a specific map from being rankable.
If this gets fixed later, **it'll probably require a leaderboard reset on affected beatmaps**. So if this is something we want to consider we should probably do it now. **This is something that is not 100% agreed upon and that needs official community feedback**.

- **Feature implementation** : Tiny droplets derandomization: This is kind of related to the issue above since it sometimes generates impossible droplets, mainly on really high CS (so it mainly happens on some ranked converts). An option should be added to control that in the editor (remove or control randomization).

- **Feature implementation** : S ranks are too easily achievable in most maps in catch, and an FC will result in the same S rank in the leaderboard. We should have a way to highlight an FC in the leaderboard (I believe it also applies to taiko).

**Comment by SecreOsu** [comment]
Adding on to the summarization for first 2 points:

> Bug fix : Notes can appear out of bounds (you can place out of bounds fruits in editor: https://github.com/ppy/osu/issues/23810), any out of bounds object will appear during gameplay. On stable, their positions are reset at x=0 or something of that effect.
> This is high priority because some ranked maps are broken because of this issue (Quoting Secre: " https://osu.ppy.sh/beatmapsets/1062804#fruits/2234731 slider at 2:00 on PH diff, several other ranked or loved maps are touched ").

The difference between stable and lazer is that in stable, if you have notes outside of the x0-512 range, they will be forced to appear on either x=0 or x=512 respectively. In lazer, this is not the case, and they can appear outside of those ranges in gameplay, but are not able to be caught. It seems that a check needs to be added to make all notes be within the borders of the playfield.

> Feature implementation : Remove object randomization for HardRock. This is the most requested and agreed upon feature change. Currently, objects receive a random offset when HR is on. This could be a fun idea but it naturally makes many patterns barely playable on HR and is a struggle for HR players.
> The randomization should still exist in classic (for compatibility or just people who want it). We have been discussing keeping the randomization feature but as a future adjustable mod.
> This only requires a few lines of code and pretty much no testing (I think?). We still need to get official community feedback for that.

I'm pretty sure the single line of code is related to "catchProcessor.HardRockOffsets", just removing this would fix what's requested afaik

**Reply by bastoo0** [reply]
> I'm pretty sure the single line of code is related to "catchProcessor.HardRockOffsets", just removing this would fix what's requested afaik

I believe it should be kept but only effective when classic mod is applied.

**Reply by SecreOsu** [reply]
yep, you're correct

**Comment by bdach** [comment]
In order:

> Notes can appear out of bounds

✅ Fixed: https://github.com/ppy/osu/pull/26158

> Remove object randomization for HardRock. This is the most requested and agreed upon feature change. Currently, objects receive a random offset when HR is on. This could be a fun idea but it naturally makes many patterns barely playable on HR and is a struggle for HR players.

❌ Not happening this year, and if it does, it will probably be behind a **hard rock** (not classic) mod setting (defaulting to randomisation on by default)

> There is an existing issue on both lazer and stable where droplets are generated in an inconsistent manner on low BPM

❌ Not happening this year for sure, not sure if ever

> This lazer scoring issue should probably be reviewed as it seems to create imbalance in some leaderboards

✅ Fixed: https://github.com/ppy/osu/pull/26405

> Inconsistent layering (see https://github.com/ppy/osu/issues/25827, also not described here: slider ticks can appear in front of normal objects).

✅ Fixed: https://github.com/ppy/osu/pull/26336

> HP is perceived as "too punishing" in catch. On high difficulties, missing an hyperdash leading to a stack will cause an inevitable fail most of the time. HP calculation should prevent failing on a singular chunk of objects, especially at high BPM.

🕒 Can be fixed later (as opposed to if it was too lenient)

> Tiny droplets Hyperdashes: This is a stable issue

❌ Not happening this year for sure, not sure if ever

> Tiny droplets derandomization

❌ Not happening this year for sure, not sure if ever

> S ranks are too easily achievable in most maps in catch, and an FC will result in the same S rank in the leaderboard. We should have a way to highlight an FC in the leaderboard (I believe it also applies to taiko).

This is a concern for all rulesets generally so it will be addressed universally

**Reply by bastoo0** [reply]
I do agree with all points except for HardRock randomization. I believe it should get further discussion and a proper poll since I think the majority of players would not want default randomization in the long run.

**Reply by SecreOsu** [reply]
> ❌ Not happening this year, and if it does, it will probably be behind a hard rock (not classic) mod setting (defaulting to randomisation on by default)

This one is very disappointing. This is a single line of code that has been a thorn in the side of every single catch HR player for the past decade. Its been being asked for for countless years, we finally have the opportunity to make it happen but it simply wont? 

This frankly *needs* to happen. Removing the randomisation from HR by default and moving it to classic is the ideal solution for every catch player.

**Reply by bdach** [reply]
It does not need to happen this year to go live with leaderboards is the point. What we need to go live is being as close to parity to stable as possible.

We can probably have the setting default to randomisation off and backpopulate existing scores to have it on even. But it _does not need to happen this year_.

**Reply by bastoo0** [reply]
> It does not need to happen this year to go live with leaderboards is the point. What we need to go live is being as close to parity to stable as possible.

Sounds fine. So older scores will get the randomization option on by default. So how do you plan to separate those from new HR scores in leaderboard? Just leaving them as they are but displaying them with an indicator?

**Reply by bdach** [reply]
Something like it yes. I believe web already shows mod settings as a tooltip. I'm not sure if we'll add support for filtering scores by mod settings but possibly that can be explored.

I guess your proposal of having CL+HR activate the offsets may also be fine but I have historically been reluctant to have mods influence the operation of other mods as it is generally messy and leads to ugly kludges. So I would prefer to have it self-contained in HR.

**Reply by bdach** [reply]
Also note that I'd say that _all_ "not happening this year" issues are issues that *may* be explored at a later point. They all appear to me to make the game *easier*, so they are generally easier things to push out to live leaderboards than a change that makes the game harder.

The only reason I am doing the hard rejection to those is that (a) they are not needed for stable parity, and (b) I was not aware of these and did not have the chance to discuss the implications of these with the rest of the team, so I do not want to promise something that I may not be able to later deliver on.

My current operating timeline is "what can I feasibly do this week before we go live".

**Reply by bastoo0** [reply]
> The only reason I am doing the hard rejection to those is that (a) they are not needed for stable parity, and (b) I was not aware of these and did not have the chance to discuss the implications of these with the rest of the team, so I do not want to promise something that I may not be able to later deliver on.

That's why I've tried to point out stuff that could potentially cause issues with leaderboards. I understand that this feedback was given really late into the year, so feel free to take what you feel is feasible or important out of it.

**Reply by Crafterdark** [reply]
Any news on the hard rock mod setting?
